### PR TITLE
Teacher dash redirect unauthorized users

### PR DESCRIPTION
**Pull request by Nokondi**
This update adds the following teacher dashboard redirects:
- Users not logged in or unauthorized to view the dashboard are redirected home
- Teachers are redirected to the appropriate course or unit page on their dashboard

## Links

[TEACH-1388](https://codedotorg.atlassian.net/browse/TEACH-1388)

## Testing story

Updated unit tests in teacher_dashboard_controller_tests to account for new redirects

## PR Checklist:

- [ ] Tests provide adequate coverage
- [ ] Privacy and Security impacts have been assessed
- [ ] Code is well-commented
- [ ] New features are translatable or updates will not break translations
- [ ] Relevant documentation has been added or updated
- [ ] User impact is well-understood and desirable
- [ ] Pull Request is labeled appropriately
- [ ] Follow-up work items (including potential tech debt) are tracked and linked

### CODE REVIEW

File: dashboard/app/controllers/courses_controller.rb [hunk 1]
```diff
@@ -30,11 +30,6 @@ def new
   end
 
   def show
-    if !params[:section_id] && current_user&.last_section_id
```
Comment (lfryemason) [inline dashboard/app/controllers/courses_controller.rb]: Can we move this removal to the other blocked PR? I don't want to change anything for the current course overview page

File: dashboard/test/controllers/teacher_dashboard_controller_test.rb [hunk 2]
```diff
@@ -11,20 +11,23 @@ class TeacherDashboardControllerTest < ActionController::TestCase
 
   test 'index: returns forbidden if no logged in user' do
```
Comment (lfryemason) [inline dashboard/test/controllers/teacher_dashboard_controller_test.rb]:
```suggestion
  test 'index: redirects to home if no logged in user' do
```

Comment (lfryemason) [reply dashboard/test/controllers/teacher_dashboard_controller_test.rb]: Same for tests below

File: dashboard/test/controllers/teacher_dashboard_controller_test.rb [hunk 3]
```diff
@@ -11,20 +11,23 @@ class TeacherDashboardControllerTest < ActionController::TestCase
...
+    assert_redirected_to "http://test.host/home"
   end
 
   test 'index: returns forbidden if requested section does not belong to teacher' do
     sign_in @section_owner
     other_teacher_section = create :section
     get :show, params: {section_id: other_teacher_section.id}
-    assert_response :forbidden
+    assert_response :redirect
+    assert_redirected_to "http://test.host/home"
```
Comment (lfryemason) [inline dashboard/test/controllers/teacher_dashboard_controller_test.rb:30]: Can we also add a test that it goes to the correct course page as well as these home redirects?

File: dashboard/app/controllers/teacher_dashboard_controller.rb [hunk 4]
```diff
@@ -1,6 +1,14 @@
 class TeacherDashboardController < ApplicationController
   load_and_authorize_resource :section
 
+  rescue_from CanCan::AccessDenied do
+    if params[:course_name]
```
Comment (lfryemason) [inline dashboard/app/controllers/teacher_dashboard_controller.rb]: While we are here, do you want to also add the unit page? Totally fine if you want to wait until later

Comment (lfryemason) [reply dashboard/app/controllers/teacher_dashboard_controller.rb]: Do we need to do anything in the routes file to add the :course_name param? I can't quite remember how we did it

File: dashboard/test/controllers/teacher_dashboard_controller_test.rb (hunk 2 above)
Comment (Nokondi) [reply dashboard/test/controllers/teacher_dashboard_controller_test.rb:12]: Can you clarify this?

File: dashboard/app/controllers/teacher_dashboard_controller.rb [hunk 5]
```diff
@@ -1,6 +1,16 @@
 class TeacherDashboardController < ApplicationController
   load_and_authorize_resource :section
 
+  rescue_from CanCan::AccessDenied do
+    if params[:course_version_name]
+      redirect_to "/courses/#{params[:course_version_name]}"
+    elsif params[:unitName]
```
Comment (lfryemason) [inline dashboard/app/controllers/teacher_dashboard_controller.rb]: We should probably match caps case for these.
```suggestion
    elsif params[:unit_name]
```

**Review by lfryemason — APPROVED** [review]
Woo! Looks great, just one tiny non-blocking nit

**Pull request merged**